File: backend/app/domain/graph.py
from pydantic import BaseModel, Field


class CytoscapeNodeData(BaseModel):
    id: str
    label: str
    type: str
    confidence: float | None = None
    sources: list[str] = Field(default_factory=list)
    search_text: str = ""


class CytoscapeNode(BaseModel):
    data: CytoscapeNodeData


class CytoscapeEdgeData(BaseModel):
    id: str
    source: str
    target: str
    label: str
    confidence: float | None = None
    search_text: str = ""


class CytoscapeEdge(BaseModel):
    data: CytoscapeEdgeData


class GraphCounts(BaseModel):
    nodes: int = Field(ge=0)
    relationships: int = Field(ge=0)
    node_types: int = Field(ge=0)
    relationship_types: int = Field(ge=0)


class GraphResponse(BaseModel):
    nodes: list[CytoscapeNode]
    edges: list[CytoscapeEdge]
    counts: GraphCounts


class NodeRelationship(BaseModel):
    id: str
    type: str
    direction: str
    other_node_id: str
    other_node_label: str
    confidence: float | None = None


class GraphNodeDetail(BaseModel):
    id: str
    name: str
    type: str
    labels: list[str] = Field(default_factory=list)
    confidence: float | None = None
    description: str | None = None
    sources: list[str] = Field(default_factory=list)
    properties: dict[str, object] = Field(default_factory=dict)
    evidence: list[dict[str, object]] = Field(default_factory=list)
    relationships: list[NodeRelationship] = Field(default_factory=list)


class CypherQueryRequest(BaseModel):
    query: str = Field(min_length=1, max_length=5_000)
    parameters: dict[str, object] = Field(default_factory=dict)


class CypherQueryResponse(BaseModel):
    rows: list[dict[str, object]]
    row_count: int = Field(ge=0)

File: frontend/src/types/graph.ts
export interface GraphNodeData {
  id: string;
  label: string;
  type: string;
  confidence: number | null;
  sources: string[];
  search_text: string;
}

export interface GraphEdgeData {
  id: string;
  source: string;
  target: string;
  label: string;
  confidence: number | null;
  search_text: string;
}

export interface GraphData {
  nodes: Array<{ data: GraphNodeData }>;
  edges: Array<{ data: GraphEdgeData }>;
  counts: { nodes: number; relationships: number; node_types: number; relationship_types: number };
}

export interface NodeRelationship {
  id: string;
  type: string;
  direction: "incoming" | "outgoing";
  other_node_id: string;
  other_node_label: string;
  confidence: number | null;
}

export interface GraphNodeDetail {
  id: string;
  name: string;
  type: string;
  labels: string[];
  confidence: number | null;
  description: string | null;
  sources: string[];
  properties: Record<string, unknown>;
  evidence: Array<Record<string, unknown>>;
  relationships: NodeRelationship[];
}

export interface CypherResult {
  rows: Array<Record<string, unknown>>;
  row_count: number;
}
